### ISSUE TITLE
training_step_end doesn't work as stated in docs

### ISSUE THREAD

**Issue by jonvet** [PyTorchLightning/pytorch-lightning#8105]
## 🐛 Bug

I'm training a model on multiple GPUs using the `dp` accelerator. I want to combine the outputs using the function `training_step_end`. However, unlike stated in the docs, in that function I don't have access to the outputs from all GPUs but just one.

## Please reproduce using the BoringModel

https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing

### To Reproduce

I can't 100% reproduce the error with BoringModel because I don't have access to (multiple) GPUs, so the behaviour of `dp` can't be tested. 
But when I run this on a 4 GPU machine, the print statement in `training_step_end` reveal that it just gets a dictionary with a logit/labels tensor of length 10, instead of 4 logit/labels tensors of length 10:

```
Epoch 10:  74%|███████▍  | 7428/10000 [01:06<00:22, 112.52it/s, loss=0.7, v_num=0]  {'logits': tensor([[-0.5810, -0.8195],
2021-06-23T15:50:47.151068998Z:         [-0.8803, -0.5356],
2021-06-23T15:50:47.151072929Z:         [-0.4559, -1.0047],
2021-06-23T15:50:47.151076585Z:         [-0.7245, -0.6628],
2021-06-23T15:50:47.151080187Z:         [-0.9753, -0.4733],
2021-06-23T15:50:47.151083827Z:         [-0.6311, -0.7593],
2021-06-23T15:50:47.151088058Z:         [-0.5547, -0.8539],
2021-06-23T15:50:47.151092906Z:         [-0.7836, -0.6102],
2021-06-23T15:50:47.151097823Z:         [-0.7722, -0.6199],
2021-06-23T15:50:47.151104739Z:         [-0.7415, -0.6470]], device='cuda:0', grad_fn=<GatherBackward>), 'labels': tensor([0, 1, 0, 1, 1, 1, 1, 0, 0, 0], device='cuda:0')}
``` 
Use following [**BoringModel**](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing) 

### Expected behavior

According to the [docs](https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html), and in particular the code snippet below coming from the docs, I would expect `batch_parts` to be a tuple containing `n` elements where `n` is the number of GPUs used. However, as can be seen in BoringModel, and what I'm getting when training on a machine with 4 GPUs, is that `batch_parts` is just a dictionary containing the output of **one** `training_step` (and not 4 training steps)

```python
def training_step(self, batch, batch_idx):
    x, y = batch
    y_hat = self.model(x)
    loss = F.cross_entropy(y_hat, y)
    pred = ...
    return {'loss': loss, 'pred': pred}

def training_step_end(self, batch_parts):
    gpu_0_prediction = batch_parts[0]['pred']
    gpu_1_prediction = batch_parts[1]['pred']

    # do something with both outputs
    return (batch_parts[0]['loss'] + batch_parts[1]['loss']) / 2

def training_epoch_end(self, training_step_outputs):
   for out in training_step_outputs:
       # do something with preds
```

### Environment

```
* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           10.2
* Packages:
	- numpy:             1.19.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.9.0+cu102
	- pytorch-lightning: 1.3.7post0
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.7.10
	- version:           #1 SMP Sat Jun 5 09:50:34 PDT 2021
```

**Comment by awaelchli** [comment]
Hi

Please have a look at the following runnable example. It requires 2 GPUs.

```python
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.plugins import DataParallelPlugin


class RandomDataset(Dataset):

    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):

    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss, "batch": batch}

    def training_step_end(self, step_outputs):
        losses = step_outputs["loss"]
        assert len(losses) == 2
        loss = losses.mean()

        batches = step_outputs["batch"]
        assert batches.shape[0] == 2
        assert batches.shape[1] == 32
        print(batches)
        return loss

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=2,
        limit_val_batches=2,
        num_sanity_val_steps=0,
        max_epochs=1,
        weights_summary=None,
        accelerator="dp",
        gpus=2,
    )
    assert isinstance(trainer.training_type_plugin, DataParallelPlugin)
    trainer.fit(model, train_dataloader=train_data)


if __name__ == '__main__':
    run()

```

**Comment by awaelchli** [comment]
Please have a look at the following example and my comments. 
It requires 2 GPUs.

```python
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.plugins import DataParallelPlugin


class RandomDataset(Dataset):

    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):

    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        assert batch.shape[0] == 4
        assert batch.shape[1] == 32
        self.log("train_loss", loss)
        return {"loss": loss, "batch": batch}

    def training_step_end(self, step_outputs):
        losses = step_outputs["loss"]
        assert losses.shape[0] == 2  # loss from both gpus
        loss = losses.mean()  # average (default if you don't implement this hook)

        batches = step_outputs["batch"]
        assert batches.shape[0] == 8  # 2 x 4, each GPU saw a batch of size 4
        assert batches.shape[1] == 32
        return loss

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=8)  # each gpu gets a batch size 4

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=2,
        limit_val_batches=2,
        num_sanity_val_steps=0,
        max_epochs=1,
        weights_summary=None,
        accelerator="dp",
        gpus=2,
    )
    assert isinstance(trainer.training_type_plugin, DataParallelPlugin)
    trainer.fit(model, train_dataloader=train_data)


if __name__ == '__main__':
    run()

```

The docs may need an update?

**Comment by tchaton** [comment]
Hey @awaelchli,

I think a doc update would be good there :)

Best,
T.C

**Comment by macwiatrak** [comment]
Thanks for the hint @awaelchli! It works well now, it would be great to update the docs.

**Comment by tchaton** [comment]
Dear @macwiatrak,

Would you mind making a PR with doc updates ?

Best,
T.C

**Comment by satishjasthi** [comment]
@tchaton 

Can I submit a PR to update the docs?
